nixos/modules/programs/z.nix

```diff
@@ -0,0 +1,109 @@
+with builtins;
+{ config, lib, pkgs, ... }:
+  let
+    l = lib; p = pkgs; t = l.types;
+
+    make-bool-options =
+      mapAttrs
+        (_: description:
+           l.mkOption
+             { type = t.bool;
+               inherit description;
+               default = true;
+             }
+        );
+
+    make-str-options =
+      mapAttrs
+        (_: description:
+           l.mkOption
+             { type = t.nullOr t.str;
+               inherit description;
+               default = null;
+             }
+        );
+
+    bool-options =
+      { resolve-symlinks = "resolve symlinks";
+
+        prompt-command =
+          ''
+          Automatically have z handle the PROMPT_COMMAND/precmd.
+
+          Note: if you override your PROMPT_COMMAND after z does, it will override the changes z makes, and z will not work. You can do something like
+
+          export PROMPT_COMMAND="my-prompt-command $PROMPT_COMMAND"
+
+          to fix this.
+          '';
+      };
+
+    str-options =
+      { cmd = "Set the name of the command (defaults: z).";
+        data = "Change the datafile (default: $HOME/.z).";
+        owner = "To allow usage when in 'sudo -s' mode";
+      };
+  in
+  { options.programs.z =
+      { enable = l.mkEnableOption "z";
+
+        max-score =
+          l.mkOption
+            { type = t.nullOr t.int;
+              description = "Lower to age entries out faster (default: 9000)";
+              default = null;
+            };
+
+        exclude-dirs =
+          l.mkOption
+            { type = t.listOf t.path;
+              description = "An array of directories to exclude";
+              default = [];
+            };
+      }
+      // make-bool-options bool-options
+      // make-str-options str-options;
+
+    config =
+      let cfg = config.programs.z; in
+      { environment =
+          { interactiveShellInit =
```

Comment:
What shells does this support? I know that there's a special fish z program, I think it would be cleaner to add it to each enabled shells' init than generally everywhere

Correct me if this is not applicable here

Reply:
> What shells does this support?

I think it tries to be as portable as possible, so maybe all POSIX shells? Looking through the codebase, I see only bash and zsh mentioned explicitly.

Reply:
I can add it to just bash's and zhs's if you think that would be most appropriate.